### DISCUSSION TITLE
Client-side middleware?

### DISCUSSION THREAD

**Discussion by entity** [nuxt/nuxt#17981] [Questions]
Hi all,

With Nuxt V3, where do we declare our middleware? With v2, we could do the following:
1. Create a file, e.g. `fathom` within the *middleware* folder.
```js
export default function(context) {
    if (!process.client) {
        return
    }
 
    window.fathom.trackPageview()
}
```

2. Reference it within our router, like so:
```js
router: {
    middleware: [
        'fathom'
    ]
}
```

3. Finally, add our script tag, for loading in analytics (as an example):
```js
// Script tags
script: [
    // Fathom Analytics
    {
        // Use your Fathom custom domain if you have one
        src: 'https://cdn.usefathom.com/script.js',
        'data-site': 'Your-Site-ID-Goes-Here',
        'defer': true
    }
]
```

How does this change with v3? As right now, analytics only track the page loaded, because there is no javascript to execute on the client.

**Comment by wobsoriano** [comment]
I think that could be in layout instead? Something like this:

```html
<script setup>
const router = useRouter();

router.beforeEach((to, from) => {
  window.fathom.trackPageview()
});
</script>

<template>
  <div>
    <slot />
  </div>
</template>
```

or maybe create a client plugin out of it?

```ts
// plugins/fathom.client.ts
import { defineNuxtPlugin } from '#app'

export default defineNuxtPlugin(nuxtApp => {
 window.fathom.trackPageview()
})
```

Edit: I checked the fathom docs and looks like the plugin example above should work.

**Reply by entity** [reply]
Oh wow, I didn’t know we did this, and this performs client side?

**Reply by entity** [reply]
It seems to stack up multiple times as each page goes by?

**Reply by wobsoriano** [reply]
added plugin example to my answer

**Reply by entity** [reply]
@wobsoriano would I enter this plugin into my config?

**Reply by wobsoriano** [reply]
> @wobsoriano would I enter this plugin into my config?

No need

**Reply by entity** [reply]
@wobsoriano this only runs on first load sadly (plugin)

**Comment by wobsoriano** [comment]
https://v3.nuxtjs.org/docs/directory-structure/middleware/
